code/Controllers/ModelAsController.php

```diff
@@ -32,30 +32,14 @@ class ModelAsController extends Controller implements NestedController {
 	 * Get the appropriate {@link ContentController} for handling a {@link SiteTree} object, link it to the object and
 	 * return it.
 	 *
+	 * @deprecated 5.0
 	 * @param SiteTree $sitetree
 	 * @param string $action
 	 * @return ContentController
 	 */
 	public static function controller_for(SiteTree $sitetree, $action = null) {
-		if ($sitetree->class == 'SilverStripe\\CMS\\Model\\SiteTree') {
-			$controller = "SilverStripe\\CMS\\Controllers\\ContentController";
-		} else {
-			$ancestry = ClassInfo::ancestry($sitetree->class);
-			while ($class = array_pop($ancestry)) {
-				if (class_exists($class . "_Controller")) {
-					break;
-				}
-			}
-			$controller = ($class !== null)
-				? "{$class}_Controller"
-				: "SilverStripe\\CMS\\Controllers\\ContentController";
-		}
-
-		if($action && class_exists($controller . '_' . ucfirst($action))) {
-			$controller = $controller . '_' . ucfirst($action);
-		}
-
-		return class_exists($controller) ? Injector::inst()->create($controller, $sitetree) : $sitetree;
+		Deprecation::notice('5.0', 'Use $sitetree->getAssociatedController() instead');
+		return $sitetree->getAssociatedController($action);
```

Comment:
I was unsure about this at first - is it better for the model to choose a controller, or for the controller to work it out itself? The more I think about it, the more I think this approach makes sense, especially given that:

1. If a controller lives in a different namespace to its “associated” page type, it’s hard to work out which controller we should use (we can’t just to `{$class}_Controller` like we do currently).
2. This approach would allow two page classes to share a controller without relying on the `{$class}_Controller` logic & inheritance